arch/arm/core/thread.c

```diff
@@ -165,3 +165,98 @@ void configure_builtin_stack_guard(struct k_thread *thread)
 #endif
 }
 #endif /* CONFIG_BUILTIN_STACK_GUARD */
+
+#if defined(CONFIG_MPU_STACK_GUARD) || defined(CONFIG_USERSPACE)
+
+#define IS_MPU_GUARD_VIOLATION(guard_start, fault_addr, stack_ptr) \
+	(fault_addr == -EINVAL) ? \
+	((fault_addr >= guard_start) && \
+	(fault_addr < (guard_start + MPU_GUARD_ALIGN_AND_SIZE)) && \
+	(stack_ptr < (guard_start + MPU_GUARD_ALIGN_AND_SIZE))) \
+	: \
+	(stack_ptr < (guard_start + MPU_GUARD_ALIGN_AND_SIZE))
+
+/**
+ * @brief Assess occurrence of current thread's stack corruption
+ *
+ * This function performs an assessment whether a memory fault (on a
+ * given memory address) is the result of stack memory corruption of
+ * the current thread.
+ *
+ * Thread stack corruption for supervisor threads or user threads in
+ * privilege mode (when User Space is supported) is reported upon an
+ * attempt to access the stack guard area (if MPU Stack Guard feature
+ * is supported). Additionally the current PSP (process stack pointer)
+ * must be pointing inside or below the guard area.
+ *
+ * Thread stack corruption for user threads in user mode is reported,
+ * if the current PSP is pointing below the start of the current
+ * thread's stack.
+ *
+ * Notes:
+ * - we assume a fully descending stack,
+ * - we assume a stacking error has occurred,
+ * - the function shall be called when handling MemManage and Bus fault,
+ *   and only if a Stacking error has been reported.
+ *
+ * If stack corruption is detected, the function returns the lowest
+ * allowed address where the Stack Pointer can safely point to, to
+ * prevent from errors when un-stacking the corrupted stack frame
+ * upon exception return.
+ *
+ * @param fault_addr memory address on which memory access violation
+ *                   has been reported. It can be invalid (-EINVAL),
+ *                   if only Stacking error has been reported.
+ * @param psp        current address the PSP points to
+ *
+ * @return The lowest allowed stack frame pointer, if error is a
+ *         thread stack corruption, otherwise return 0.
+ */
+u32_t evaluate_current_thread_stack_corruption(const u32_t fault_addr,
```

Comment:
this function isn't properly namespaced, needs a z_ prefix
The name of this function is excessively long.

Reply:
Fixed. Thanks.